pyquil/parameters.py

```diff
@@ -0,0 +1,273 @@
+        self.op1 = op1
+        self.op2 = op2
+
+
+class Add(BinaryExp):
+    operator = '+'
+    precedence = 1
+    associates = 'both'
+
+    def __init__(self, op1, op2):
+        super(Add, self).__init__(op1, op2)
+
+
+class Sub(BinaryExp):
+    operator = '-'
+    precedence = 1
+    associates = 'left'
+
+    def __init__(self, op1, op2):
+        super(Sub, self).__init__(op1, op2)
+
+
+class Mul(BinaryExp):
+    operator = '*'
+    precedence = 2
+    associates = 'both'
+
+    def __init__(self, op1, op2):
+        super(Mul, self).__init__(op1, op2)
+
+
+class Div(BinaryExp):
+    operator = '/'
+    precedence = 2
+    associates = 'left'
+
+    def __init__(self, op1, op2):
+        super(Div, self).__init__(op1, op2)
+
+
+class Pow(BinaryExp):
+    operator = '^'
+    precedence = 3
+    associates = 'right'
+
+    def __init__(self, op1, op2):
+        super(Pow, self).__init__(op1, op2)
+
+
+def _expression_to_string(expression):
+    """
+    Recursively converts an expression to a string taking into account precedence and associativity for placing
+    parenthesis
+
+    :param Expression expression: expression involving parameters
+    :return: string such as '%x*(%y-4)'
+    :rtype: str
+    """
+    if isinstance(expression, BinaryExp):
+        left = _expression_to_string(expression.op1)
+        if isinstance(expression.op1, BinaryExp) and not (
+                expression.op1.precedence > expression.precedence or
+                expression.op1.precedence == expression.precedence and
+                expression.associates in ('left', 'both')):
+            left = '(' + left + ')'
+
+        right = _expression_to_string(expression.op2)
+        if isinstance(expression.op2, BinaryExp) and not (
+                expression.precedence < expression.op2.precedence or
+                expression.precedence == expression.op2.precedence and
+                expression.associates in ('right', 'both')):
+            right = '(' + right + ')'
+
+        return left + expression.operator + right
+    elif isinstance(expression, Function):
+        return expression.name + '(' + _expression_to_string(expression.expression) + ')'
+    elif isinstance(expression, Parameter):
+        return str(expression)
+    else:
+        return format_parameter(expression)
+
+
+def _contained_parameters(expression):
+    """
+    Determine which parameters are contained in this expression.
+
+    :param Expression expression: expression involving parameters
+    :return: set of parameters contained in this expression
+    :rtype: set
+    """
+    if isinstance(expression, BinaryExp):
+        return _contained_parameters(expression.op1) | _contained_parameters(expression.op2)
+    elif isinstance(expression, Function):
+        return _contained_parameters(expression.expression)
+    elif isinstance(expression, Parameter):
+        return {expression}
+    else:
+        return set()
+
+
+def _check_for_pi(element):
```

Comment:
How about offering a `pi` symbol that people can use in their expressions?

Reply:
Going to punt on this for now but it's a good idea to get better pi printing in more cases